### DISCUSSION TITLE
How to install user libraries in OM 1.20?

### DISCUSSION THREAD

**Discussion by ceraolo** [OpenModelica/OpenModelica#9813] [Q&A]
OM 1.20 uses well-documented package manager to install libraries from the OpenModelica site.
But how to install user libraries? It seems that the documentation does not say anything about this.

With older OM versions I just copied my personal library in "omlibrary". For instance:
C:\OM\OpenModelica 1.18.0-64bit\lib\omlibrary
and worked very well.

Now, this seems not possible anymore.  How to proceed to install user libraries with OM 1.20??

Thank all.

**Comment by adeas31** [comment]
Why do you want to load user library in this way? This will make user library read-only.

I am not entirely sure but you can copy your library to `%APPDATA%/.openmodelica/libraries`. I don't know what will happen with this library when the package manager will try to update libraries.

**Reply by ceraolo** [reply]
> Why do you want to load user library in this way? This will make user library read-only.

**Example 1:**
I have is a library (EHTPlib) developed over years, that I use for teaching. For me its quality is good enough to preserve it from undue modifications, i.e. to make it read-only, similar to system libraries.
Sooner or later I will propose this library for consideration by OM to be included as a supported library. On Github it has been forked 8 times: not much but significant.
Moreover, I find very nice that my EHPTlib loads automatically because of a uses annotation. It makes the user workflow smoother.

**Examples 2 and 3:**
I use PowerGrids and, from time to time, PowerSystems. Both libraries are written in MSL 3.2.3; but I use submodels from them in some of my MSL 4.0 models. Therefore I've converted PowerGrids (and want to convert PowerSystems) into 4.0, and use these personalised libraries as local system libraries.
> 
> I am not entirely sure but you can copy your library to `%APPDATA%/.openmodelica/libraries`. I don't know what will happen with this library when the package manager will try to update libraries.

I'll try and report.

**Reply by ceraolo** [reply]
> Why do you want to load user library in this way? This will make user library read-only.
> 
> I am not entirely sure but you can copy your library to `%APPDATA%/.openmodelica/libraries`. I don't know what will happen with this library when the package manager will try to update libraries.

I checked and this suggestion for me works very well.

**Comment by casella** [comment]
@ceraolo points out a quite reasonable user requirement. Currently, we only manage read-only system libraries if they are a) open source and b) listed in OSMC's config.json scripts. One could, e.g. for teaching purposes, want to have other read-only system libraries that for some reason fall outside these categories.

I think this could be done very easily, provided that putting extra packages in the .openmodelica/libraries directory does not have unintended consequences on the package manager. I don't think that is the case, @sjoelund can you please confirm?

If that is possible, we could add a menu item File | Manage Libraries | Install Local Library, where users can select a directory or .mo file that contains a library in their file system, which will then be copied to .openmodelica/libraries.

This would also very conveniently address the needs for off-line or behind-firewall corporate environments.

@adeas31 what do you think?

**Reply by adeas31** [reply]
Sure, copying stuff shouldn't be a problem.

Btw there is another alternative, just add another path to `MODELICAPATH` that points to the library OR perhaps add a path to `MODELICAPATH` like `user-libraries` and then add all user libraries there. That should also work.

**Comment by mahge** [comment]
I have done this multiple times and I have not experienced any issues so far. The one risk you have is that you might forget you added it there and delete the whole thing when the package manager acts up for other reasons.

**Reply by casella** [reply]
Well, at the moment AFAIK the package manager does not delete anything, so I am not worried by this eventuality

**Reply by mahge** [reply]
I meant deleting manually. I used to delete packages and reinstall them from the cache when things got mixed up (with uses and versions or index.json IIRC). However, I have not seen those issues in a long time now. It should be even safer.

**Comment by sjoelund** [comment]
Having an option to copy a package to the home directory has the advantage that you can add the JSON metadata so the package manager knows it was manually installed and the uses/etc of the package.

Note that adding directories to the Modelica path disables the package manager at the moment.

**Reply by adeas31** [reply]
So then we need an API for it so that package manager can update the JSON accordingly.

**Reply by casella** [reply]
> Note that adding directories to the Modelica path disables the package manager at the moment.

I guess you mean changing the MODELICAPATH string, not actually adding directories to the paths pointed by it, right?

So I see we have three options to install user libraries:
1. have OMEdit copy them to .opemodelica/libraries
2. have OMEdit copy them to .opemodelica/user-libraries and add that library to the MODELICAPATH permanently (without breaking the package manager)
3. have OMEdit copy them to .openmodelica/libraries and update the JSON file with appropriate metadata

The first seems very easy, but the package manager doesn't know about them
The second is a bit more complicated, could break the package manager, which still doesn't know about them
The third is two bits more complicated but package manager is aware - not clear however how to merge the local info when you update index.json from the OSMC server
